parquet-avro/src/test/java/org/apache/parquet/avro/TestWriteVariant.java

```diff
@@ -0,0 +1,518 @@
+    }
+  }
+
+  @Test
+  public void testShreddedValues() throws IOException {
+    for (Variant v : VARIANTS) {
+      GenericRecord record = createRecord(1, v);
+      MessageType writeSchema = shreddingSchema(v);
+      TestSchema testSchema = new TestSchema(writeSchema, READ_SCHEMA);
+
+      GenericRecord actual = writeAndRead(testSchema, record);
+      assertEquals(record.get(0), actual.get(0));
+      Variant actualV = new Variant((ByteBuffer) ((GenericRecord) actual.get(1)).get(1), (ByteBuffer)
+          ((GenericRecord) actual.get(1)).get(0));
+      AvroTestUtil.assertEquivalent(v, actualV);
+    }
+  }
+
+  @Test
+  public void testMixedShredding() throws IOException {
+    for (Variant valueForSchema : VARIANTS) {
+      List<GenericRecord> expected = new ArrayList<>();
+      for (int i = 0; i < VARIANTS.length; i++) {
+        expected.add(createRecord(i, VARIANTS[i]));
+      }
+
+      MessageType writeSchema = shreddingSchema(valueForSchema);
+      TestSchema testSchema = new TestSchema(writeSchema, READ_SCHEMA);
+
+      List<GenericRecord> actual = writeAndRead(testSchema, expected);
+      assertEquals(actual.size(), VARIANTS.length);
+      for (int i = 0; i < VARIANTS.length; i++) {
+        Variant actualV =
+            new Variant((ByteBuffer) ((GenericRecord) actual.get(i).get(1)).get(1), (ByteBuffer)
+                ((GenericRecord) actual.get(i).get(1)).get(0));
+        AvroTestUtil.assertEquivalent(VARIANTS[i], actualV);
+      }
+    }
+  }
+
+  // Write schema contains the full shredding schema. Read schema should just be a value/metadata pair.
+  private static class TestSchema {
+    MessageType writeSchema;
+    MessageType readSchema;
+
+    TestSchema(MessageType writeSchema, MessageType readSchema) {
+      this.writeSchema = writeSchema;
+      this.readSchema = readSchema;
+    }
+  }
+
+  /**
+   * This is a custom Parquet writer builder that injects a specific Parquet schema and then uses
+   * the Avro object model. This ensures that the Parquet file's schema is exactly what was passed.
+   */
+  private static class TestWriterBuilder extends ParquetWriter.Builder<GenericRecord, TestWriterBuilder> {
+    private MessageType schema = null;
+
+    protected TestWriterBuilder(Path path) {
+      super(path);
+    }
+
+    TestWriterBuilder withFileType(MessageType schema) {
+      this.schema = schema;
+      return self();
+    }
+
+    @Override
+    protected TestWriterBuilder self() {
+      return this;
+    }
+
+    @Override
+    protected WriteSupport<GenericRecord> getWriteSupport(Configuration conf) {
+      return new AvroWriteSupport<>(schema, new AvroSchemaConverter().convert(schema), GenericData.get());
+    }
+  }
+
+  GenericRecord writeAndRead(TestSchema testSchema, GenericRecord record) throws IOException {
+    List<GenericRecord> result = writeAndRead(testSchema, Arrays.asList(record));
+    assert (result.size() == 1);
```

Comment:
This isn't the same as a test assertion. In tests, prefer using `assertEquals(1, result.size())`